### ISSUE TITLE
cookie named \"_fz_http_key\" exceeds maximum size of 4096 bytes

### ISSUE THREAD

**Issue by T-baby** [CloudFire-LLC/cloudfire#1621]
Request: GET /auth/oidc/gate/callback/?code=fa9139cf9de508e0e16f&state=F7cC96pgwIjSuIIb
** (exit) an exception was raised:
    ** (Protocol.UndefinedError) protocol Phoenix.HTML.Safe not implemented for %Plug.Conn.CookieOverflowError{message: "cookie named \"_fz_http_key\" exceeds maximum size of 4096 bytes"} of type Plug.Conn.CookieOverflowError (a struct). This protocol is implemented for the following type(s): Atom, BitString, Date, DateTime, Decimal, Float, FzHttp.Types.IPPort, Integer, List, NaiveDateTime, Phoenix.HTML.Form, Phoenix.LiveComponent.CID, Phoenix.LiveView.Component, Phoenix.LiveView.Comprehension, Phoenix.LiveView.JS, Phoenix.LiveView.Rendered, Postgrex.INET, Time, Tuple, URI
        (phoenix_html 3.3.1) lib/phoenix_html/safe.ex:1: Phoenix.HTML.Safe.impl_for!/1
        (phoenix_html 3.3.1) lib/phoenix_html/safe.ex:15: Phoenix.HTML.Safe.to_iodata/1
        (phoenix 1.7.2) lib/phoenix/controller.ex:1010: anonymous fn/5 in Phoenix.Controller.template_render_to_iodata/4
        (telemetry 1.2.1) /app/deps/telemetry/src/telemetry.erl:321: :telemetry.span/3
        (phoenix 1.7.2) lib/phoenix/controller.ex:976: Phoenix.Controller.render_and_send/4
        (phoenix 1.7.2) lib/phoenix/endpoint/render_errors.ex:84: Phoenix.Endpoint.RenderErrors.instrument_render_and_send/5
        (phoenix 1.7.2) lib/phoenix/endpoint/render_errors.ex:62: Phoenix.Endpoint.RenderErrors.__catch__/5
        (plug_cowboy 2.6.1) lib/plug/cowboy/handler.ex:11: Plug.Cowboy.Handler.init/2

**Comment by T-baby** [comment]
How can i fix it?

**Comment by jamilbk** [comment]
Hi there, you'll need to use the `PHOENIX_HTTP_PROTOCOL_OPTIONS` env var to increase the max header size from 4096. See https://www.firezone.dev/docs/reference/env-vars/.

tl;dr: try this:

```
PHOENIX_HTTP_PROTOCOL_OPTIONS={"max_header_value_length":8192}
```

**Comment by T-baby** [comment]
Thanks for your reply, but I have added the `PHOENIX_HTTP_PROTOCOL_OPTIONS` you gave to the.env file and the same exception will still occur.

<img width="537" alt="image" src="https://github.com/firezone/firezone/assets/7859785/19202c33-424d-4812-9441-2203aa8d4fb5">

**Comment by T-baby** [comment]
I'm not sure if using http2 is relevant.

**Comment by T-baby** [comment]
Can you help me with another question first?
https://github.com/firezone/firezone/issues/1621

**Comment by T-baby** [comment]
I tried to enable SSO with zitadel,  but it also occurred the same exception.

**Comment by boppy** [comment]
Sorry, I've got no fix to provide, but an insight: Browsers "only" allow 4096 Byte cookies to be stored.

**Comment by Kuppit** [comment]
Hello,
I have exactly the same issue; it is no longer possible to connect to the interface
It's a rather blocking bug. We are using Keycloak for OIDC authentication.

**Comment by jamilbk** [comment]
Hmmm, we haven't been able to reproduce this -- there must be some parameter(s) being saved in the session that happens to be very long. Could you @Kuppit @T-baby report the number of characters you have for the following:

- Affected user's email address
- OIDC config ID
- OIDC discovery document URI
- OIDC redirect URI

Some versions of these are saved in the session -- if they're too large it could be causing the issue.

**Comment by Kuppit** [comment]
- Affected user's email address: 24 characters (xxxxxx.xxxxxx@xxxxxx.xxx)
- OIDC config ID: 8 characters
- OIDC discovery document URI: 69 characters (https://xxx.xxxxxx.xxx/realms/xxxxxx/.well-known/openid-configuration)
- OIDC redirect URI: 51 characters (https://xxx.xxxxxx.xxx/auth/oidc/xxxxxxxx/callback/)

Considering that we have been using Firezone for over 3 months, this is the first time I have seen this issue. There have been no configuration changes, only an update a few weeks ago, but it was working fine after the update.

**Comment by jamilbk** [comment]
@Kuppit That's strange, yeah it's a tough one to nail down. None of your parameters look obscenely long.

Does this happen consistently? I.e. If you clear cookies (both Firezone and Keycloak) and re-authenticate, does it still happen?

**Comment by Kuppit** [comment]
Yes it happens with each test, with different browsers, different users and even by deleting the cookies and local storage of the 2 sites.

**Comment by jamilbk** [comment]
@Kuppit We store the OIDC ID token from the successful authentication in the cookie. Have you recently added fields/metadata to this token in Keycloak or changed any of the settings there? Are you able to post a sample ID token here so we can check its length?

If that's the culprit, the short-term workaround would be to minimize the size of your ID token in Keycloak. The longer term fix won't be trivial in Firezone -- we'll need to resort to persisting these tokens in the DB since they can grow potentially unbounded.

cc @AndrewDryga for context since you're working on auth for 1.0

**Comment by Kuppit** [comment]
Indeed, upon reflection, we added several roles to the users for another tool, and we send the roles by default. I have just made the roles optional for sending, and the connection to Firezone is working again.

Thank you

**Comment by jamilbk** [comment]
The root issue should be resolved in a future release, but glad we were able to resolve.
